### PR TITLE
Add description for the append constraint in xpath constraint for access rule

### PR DESCRIPTION
**Pull request by fazzdev**

### CODE REVIEW

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md [hunk 1]
```diff
@@ -106,10 +106,16 @@ An [XPath constraint](/refguide/xpath-constraints/) can be used to constrain the
 
 {{< figure src="/attachments/refguide/modeling/domain-model/entities/access-rules/access-rule-xpath-tab.png" >}}
```
Comment (paulmendix) [inline content/en/docs/refguide/modeling/domain-model/entities/access-rules.md:108]: This is definitely not your fault, as you are trying to improve this documentation, but I find reading this existing section super difficult. To me it feels really chaotic.

Adding more information here makes it extra complex. 

# Proposal


I propose we add two separate sections:

`#### 2.4.1 Owner` and `#### 2.4.2 Path to user...`

Then the Owner section can be ⬇

---


The  **Owner** button adds an XPath constraint so the access rule is only applied to the current user. 

This constraint is only valid when the **Store 'owner'** checkbox in the **System members** section is checked.

image

---

Note that I initially really liked that you added the actual XPath constraint, but when I see it as a user I want to fully understand it, and then we get to hidden one-way navigable associations, so then I changed my mind. Maybe better to not show it.

---


Then the 'Path to user... ' section can be ⬇

---




The **Path to user...** button is useful to create an XPath constraint to check if the entity is (indirectly) connected to the current user. When you press **Path to user...** you can select a path to an entity that is derived from **System.User** and converts that into an XPath constraint for the access rule. This means the access rule is only applied when the user path you selected is equal to the current user.


---

Note that you mention `We provided 2 helper buttons`. I don't remember seeing `We` in the documentation.

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md [hunk 2]
```diff
@@ -106,10 +106,23 @@ An [XPath constraint](/refguide/xpath-constraints/) can be used to constrain the
...
+[System.owner='[%CurrentUser%]']
+```
+
+This constraint is only valid when the [Store 'owner'](/refguide/entities/#343-store-owner) checkbox in the System members section is checked.
+
+#### 2.4.2
+
+The **Path to user...** button is useful to create an XPath constraint to check if the entity is (indirectly) associated to the current user. When you press **Path to user...** you can select a path to an entity that is a `System.User` or a specialization of `System.User` and converts that into an XPath constraint for the access rule. This means the access rule is only applied when the user path you selected is equal to the current user.
+
 For example, the **Customer** entity is a specialization of the **User** entity. The **Order** entity is associated to the **Customer** entity.
```
Comment (paulmendix) [inline content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: I think this example belongs to section 2.4 and not to 2.4.2

Comment (fazzdev) [reply content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: Yes it is.

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md (hunk 1 above)
Comment (fazzdev) [reply content/en/docs/refguide/modeling/domain-model/entities/access-rules.md:108]: Done.

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md (hunk 2 above)
Comment (paulmendix) [reply content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: Then I think the sentence `This is accomplished by using the following XPath constraint in the access rule of the **Order** entity:` needs to be changed so you describe how to use the 'Path to user' button. Now it gives the impression you have to write the XPath constraint yourself.

**Review by MarkvanMents — APPROVED** [review]
Hi @fazzdev and @paulmendix 
I've made a few proofreading changes. Hope they make sense.
As far as language is concerned, I'm approving this PR.
Please can you review and make sure it is still accurate and clear.
Thanks
Mark

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md [hunk 3]
```diff
@@ -102,25 +102,43 @@ If a user cannot view the value of an attribute because of security constraints,
...
+
+#### 2.4.2 Path to User
+
+The **Path to user...** button allows you to create an XPath constraint to check that the entity is associated (directly or indirectly) with the current user. When you click **Path to user...**, you can select a path to an entity that is a `System.User` or a specialization of `System.User`. This is converted into an XPath constraint for the access rule. This means that the rule is only applied when the user path you selected is associated with the current user.
+
+#### 2.4.3 Example
+
+Assume that the **Customer** entity is a specialization of the **User** entity. The **Order** entity is associated to the **Customer** entity via the **Order_Customer** association.
 
 A logged-in customer is allowed to view personal orders, but is not allowed to view the orders of other customers. This is accomplished by using the following XPath constraint in the access rule of the **Order** entity:
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: @paulmendix - did you mean something like:
```suggestion
A logged-in customer is allowed to view personal orders, but is not allowed to view the orders of other customers. This is accomplished by using the following XPath constraint in the access rule of the **Order** entity, which you can add using the **Path to user...** button:
```

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md (hunk 2 above)
Comment (fazzdev) [reply content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: Indeed I had an attempt to refactor "For example, ..." but I ended up with more complex explanation. I choose to revert back to how it was without realizing that I re-introduced the old sentences. I will make another attempt to rephrase the example part.

Comment (fazzdev) [reply content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: I think I have it now. Can you check?

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md (hunk 3 above)
Comment (fazzdev) [reply content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: I think I have it now. Can you check if it is more readable?

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md (hunk 1 above)
Comment (paulmendix) [reply content/en/docs/refguide/modeling/domain-model/entities/access-rules.md:108]: Looks good to me

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md (hunk 2 above)
Comment (paulmendix) [reply content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: Looks good to me

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md (hunk 3 above)
Comment (paulmendix) [reply content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: Looks good to me

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md [hunk 4]
```diff
@@ -102,25 +102,42 @@ If a user cannot view the value of an attribute because of security constraints,
...
-For example, the **Customer** entity is a specialization of the **User** entity. The **Order** entity is associated to the **Customer** entity.
+{{% alert color="warning" %}}
+XPath constraints can only be applied to persistable entities as they are applied by the database. Defining XPath constraints for non-persistable entities results in consistency errors.
+{{% /alert %}}
+
+There are two constraints that can be appended easily with a single button click. 
+
+#### 2.4.1 Owner
+
+The **Owner** button adds an XPath constraint so the access rule is only applied to the current user.
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: Would it be clearer to say:
```suggestion
The **Owner** button adds an XPath constraint so the access rule is only applied if the object owner is the current user.
```

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md [hunk 5]
```diff
@@ -102,25 +102,42 @@ If a user cannot view the value of an attribute because of security constraints,
...
+[System.owner='[%CurrentUser%]']
+```
 
-A logged-in customer is allowed to view personal orders, but is not allowed to view the orders of other customers. This is accomplished by using the following XPath constraint in the access rule of the **Order** entity:
+This constraint is only valid when the [Store 'owner'](/refguide/entities/#store-owner) checkbox in the **System members** section of the entity properties is checked.
 
-```java
+#### 2.4.2 Path to User
+
+The **Path to user...** button adds an XPath constraint so the access rule is only applied when the associated object (directly or indirectly) is the current user. When you click **Path to user...**, you can select a path to an associated entity that is either a `System.User` or a specialization of `System.User`. This is then converted into an XPath constraint for the access rule.
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: Perhaps clarify that it is a *User* object?
```suggestion
The **Path to user...** button adds an XPath constraint so the access rule is only applied when the User object which is associated (directly or indirectly) is the current user. When you click **Path to user...**, you can select a path to an associated entity that is either a `System.User` or a specialization of `System.User`. This is then converted into an XPath constraint for the access rule.
```

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md [hunk 6]
```diff
@@ -102,25 +102,42 @@ If a user cannot view the value of an attribute because of security constraints,
...
+
+```java {linenos=false}
 [Module.Order_Customer = '[%CurrentUser%]']
 ```
 
-{{< figure src="/attachments/refguide/modeling/domain-model/entities/access-rules/access-rule-order-xpath.png" >}}
+As an example:
+1. Assume that the **Customer** entity is a specialization of the **User** entity. The **Order** entity is associated to the **Customer** entity via the **Order_Customer** association.
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]:
```suggestion
1. Assume that the **Customer** entity is a specialization of the **User** entity. The **Order** entity is associated with the **Customer** entity via the **Order_Customer** association.
```

**Review by MarkvanMents — COMMENTED** [review]
Looks great.
I have a couple of suggestions, but if you are happy I can publish this immediately.
Thanks for all your hard work to explain a topic which is difficult for customers (security is always hard to explain well) and generate the PR.
Mark

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md (hunk 4 above)
Comment (fazzdev) [reply content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: Agree.

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md (hunk 6 above)
Comment (fazzdev) [reply content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: Done.

File: content/en/docs/refguide/modeling/domain-model/entities/access-rules.md (hunk 5 above)
Comment (fazzdev) [reply content/en/docs/refguide/modeling/domain-model/entities/access-rules.md]: Done.

**Pull request merged**